File: rtmb_java/src/com/ibm/realtime/rtmb/tests/util/DiskIOReadUtils.java
/*******************************************************************************
 * Copyright (c) 2008 International Business Machines Corp.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Ajith Ramanath            - initial API and implementation
 *    Radhakrishnan Thangamuthu - initial API and implementation
 *    Mike Fulton               - initial API and implementation
 *******************************************************************************/
 
package com.ibm.realtime.rtmb.tests.util;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

import com.ibm.realtime.rtmb.RTMB;

public class DiskIOReadUtils {
	private FileInputStream fInputStream;
	private int readBufSize;
	private File tmpFile;

	/**
	 * Constructor
	 * @param rbsize Read buffer size
	 */
	public DiskIOReadUtils(int rbsize) {
		this.readBufSize = rbsize;
		try {
			this.tmpFile = File.createTempFile(RTMB.benchmarkName(), null);
		} catch (IOException e) {
			e.printStackTrace();
		}
		try {
			this.fInputStream = new FileInputStream(tmpFile);
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}
		setupFileForRead();
	}

	private void setupFileForRead() {
		/*
		 * Fill the file with ascii text to read later
		 * during the test.
		 */
		byte [] wrBuffer = new byte [readBufSize];
		FileOutputStream fOutputStream = null;
		for (int i = 0; i < readBufSize; i++) {
			wrBuffer[i] = 'x';
		}
		try {
			fOutputStream = new FileOutputStream(tmpFile);
			fOutputStream.write(wrBuffer);
			fOutputStream.close();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public void doRead(int loops) {
		byte [] byteArray = new byte[readBufSize];
		for (int i = 0; i < loops; i++) {
			try {
				fInputStream.read(byteArray);
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	public void cleanup () {
		try {
			fInputStream.close();
			tmpFile.delete();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}

File: rtmb_cpp/src/testcase/util/DiskIOReadUtils.cpp
/*******************************************************************************
 * Copyright (c) 2008 International Business Machines Corp.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Ajith Ramanath            - initial API and implementation
 *    Radhakrishnan Thangamuthu - initial API and implementation
 *    Mike Fulton               - initial API and implementation
 *******************************************************************************/
 
#include <DiskIOReadUtils.h>
#include <string.h>
#include <stdlib.h>
#include <iostream>
#include <assert.h>
#include <sys/types.h>
#include <unistd.h>
#include <sys/statvfs.h>
using namespace std;

#define FILE_SIZE 100
static char filename[FILE_SIZE];

DiskIOReadUtils::DiskIOReadUtils(int rbSize_, int num) {
   _readBufSize = rbSize_;
   _numberOfBlocks = num;

   _readBuf = (char *) malloc(_readBufSize * num);
   if (_readBuf != NULL) {
      memset(_readBuf, 0, _readBufSize * num);
   } else {
      abort();
   }

   pid_t pid = getpid();
   sprintf(filename, "/tmp/RTMB_disk_io_%lld", (long long) pid);
   _fp = fopen(filename, "w");
   assert(_fp != NULL);

   setupFileForRead();
   fclose(_fp);
}

void DiskIOReadUtils::setupFileForRead() {
   int i = 0;
   char *str = (char *) malloc(_readBufSize);

   if (str != NULL) {
      memset(str, 0, _readBufSize);
      for (i = 0; i < _readBufSize; i++) {
         str[i] = 'x';
      }

      for (i = 0; i < _numberOfBlocks; i++) {
         int n = fwrite(str, sizeof(char), _readBufSize, _fp);

         if (n != _readBufSize) {
            perror("setupFileForRead");
            abort();
         }
      }

      free(str);
      fflush(_fp);
   } else {
      perror("malloc:");
      abort();
   }
}

void DiskIOReadUtils::openFileForRead() {
   _fp = fopen(filename, "r");
   assert(_fp != NULL);
   setbuf(_fp, NULL);
}

void DiskIOReadUtils::closeFile() {
   if (fclose(_fp) != 0) {
      perror("fclose");
   }

   _fp = NULL;
}

void DiskIOReadUtils::cleanup() {
   if (_fp != NULL) {
      if (fclose(_fp) != 0) {
         perror("cleanup");
      }
   }

   unlink(filename);
   free(_readBuf);
}

int DiskIOReadUtils::getBlockSize() {
   struct statvfs buf;
   int rc;
   const char *path = "/tmp";

   rc = statvfs(path, &buf);

   if (rc != 0) {
      abort();
   }

   return buf.f_bsize;
}
